### PR TITLE
HARMONY-1164: Fix inconsistent job status

### PR DESCRIPTION
**Pull request by chris-durbin**
## Jira Issue ID
HARMONY-1164

## Description
We've been seeing some inconsistent job status when several work items complete work for a job around the same time. Sometimes the progress percent would go down and in some instances the status would go back to running when the request should have been successful. When performing a workload run with just a single request going at a time with a random number of granules between 1 and 97 for a harmony-service-example request I would consistently see this error condition where all work items were successful, but the job was still in the running state. I'd usually see it within just a couple of requests from the start of running and never made it past 10 requests.

After this change to make sure the job is locked when updating the status after a work item completes I've been able to run a workload run in my own sandbox for hours without issue (more than 250 requests).

Note that on a couple of occasions I've seen a case where a work item stays in the running state and the job stays running. While that needs to be figured out and fixed there is a separate ticket for handling it. This ticket is just to make sure the job status is consistent with the work items.

## Local Test Steps
Testing locally is difficult because you'd need to have many granules processed in parallel. I would recommend deploying to a sandbox environment where we run 50 in parallel.

Run a workload run with just the harmony-service-example with a random number of granules:
`HTTP_PROXY=socks5h://localhost:8080 locust --tags random`

In the locust UI choose to run just 1 worker at a time (for maximum contention on work items updating for the same job and also so the run will stop as soon as one job is hung in the running state). Make sure that the requests continue to go through successfully, or if things do get stuck make sure that not all of the work items are in the successful state for that job.

## PR Acceptance Checklist
* [X] Acceptance criteria met
* ~[ ] Tests added/updated (if needed) and passing~
* ~[ ] Documentation updated (if needed) ~

### CODE REVIEW

File: app/models/job.ts [hunk 1]
```diff
@@ -237,14 +237,29 @@ export class Job extends Record implements JobRecord {
...
   */
-  static async byJobID(transaction: Transaction, jobID: string): Promise<Job | null> {
-    const jobList = await this.queryAll(transaction, { where: { jobID } }, true, 0, 1);
-    return jobList.data.shift();
+  static async byJobID(
+    transaction: Transaction, jobID: string, getLinks = true, lock = false,
+  ): Promise<Job | null> {
+
+    let query = transaction('jobs').select().where({ jobID });
+    if (lock) {
```
Comment (chris-durbin) [inline app/models/job.ts:253]: This is really the key change for the PR - Lock the row in the jobs table when necessary. We will set lock to true as part of updating the job status in a work item update.

File: app/models/job.ts [hunk 2]
```diff
@@ -237,14 +237,29 @@ export class Job extends Record implements JobRecord {
...
+      query = query.forUpdate();
+    }
+
+    const result = await query;
+    const job = new Job(result[0]);
+    if (getLinks) {
```
Comment (chris-durbin) [inline app/models/job.ts:259]: Optimization - only load the job links if they are needed from the context of where this function is called.

File: fixtures/cmr.uat.earthdata.nasa.gov-443/165099102895369026 [hunk 3]
```diff
@@ -0,0 +1,25 @@
+POST /search/clear-scroll
```
Comment (chris-durbin) [inline fixtures/cmr.uat.earthdata.nasa.gov-443/165099102895369026:1]: I'm not sure where this fixture came from, but it kept getting generated for me. Running mocha now I don't see any new fixtures generated.

**Review by vinnyinverso — APPROVED** [review]
I like the bonus optimization, nice job.

**Review by indiejames — APPROVED** [review]

**Pull request merged**